Converts a CRTF region string to Shape class.

sushobhana opened the pull request:
This is ready to be merged
Todos :
- [x] write docstrings
- [x] Modify shape class to properly hold CRTF regions.
- [x] write documentation
- [x] write better error messages.
- [x] write tests.

keflavich commented on regions/io/crtf/read.py:
```diff
@@ -0,0 +1,334 @@
+
+        else:
+            self._raise_error("Not a valid CRTF line '{0}'.".format(line))
+            return
```
no need for a return after a raise

keflavich replied:
ah, I see, it is necessary because you can set it to 'warn'.  OK.

keflavich commented on regions/io/crtf/read.py:
```diff
@@ -0,0 +1,334 @@
+            self._raise_error("Not a valid CRTF line '{0}'.".format(line))
+            return
+
+    def _raise_error(self, msg):
```
since this is repeated code, perhaps it belongs in a parent class?

sushobhana commented:
@keflavich Thanks!! I will change them in the next commit..

sushobhana commented on regions/io/core.py:
```diff
@@ -0,0 +1,174 @@
+                msg = 'Skipping elliptical annulus {}'.format(shape)
+                warn(msg, DS9RegionParserWarning)
+                continue
+            log.debug(shape)
```
these were present before. Should I remove all the debug statements?? (There are lot of these in `regions/io/ds9/*` codes)

keflavich replied:
we can keep them for now, but we should aim to remove them eventually...

sushobhana commented:
W.r.t. the last commit , I have done some changes in the __init__.py of the subpackages, and added some imports in io/core.py . This results to an import error in tests. Can you tell me why is it happening and how can it be resolved? @keflavich

sushobhana commented:
W.r.t. the last commit , I have done some changes in the init.py of the subpackages, and added some imports in io/core.py . This results to an import error in tests. Can you tell me why is it happening and how can it be resolved? @keflavich

keflavich commented:
@sushobhana, could you add an update to your checklist?  What is the next step for test writing?  Are there parts you need help with?

sushobhana commented on regions/io/crtf/test/data/CRTFgeneral.reg:
```diff
@@ -0,0 +1,20 @@
+-ellipse[[12deg, +45deg], [0.25deg, 1.34deg], 45rad], range=[1.420GHz, 1.421GHz], corr=[Q], color=green, label='Removed this'
+
+# A diamond marker, in J2000 coordinates:
+symbol[[32.1423deg, 12.1412deg], D], linewidth=2, coord=J2000, symsize=2
```
@keflavich This is the only official sample I am having now. 
Though I can come up with basic test cases of my own.
Like I cannot be sure about the reliability of the Parser until we can have some practical CRTF files. (Like the dec representation as `-32.6.7.8` as we discussed earlier.)

keflavich replied:
I'm starting to search out more examples.  But one straightforward and important thing to do is create round-trip tests to and from ds9 regions.  So, for example, take a ds9 region file, convert each region to its CRTF equivalent, then read it back and convert to ds9 and make sure it's the same.

sushobhana replied:
I still haven't made methods for conversion between CRTF and DS9 . I will make sure I complete that and perform round tests.

sushobhana commented:
@keflavich  what do you think about storing meta attribute in the region and checking their validity.(sixth in the checklist) Now that we will also be handling spectral dimensions of the regions, I think we should think of a really good approach . (What about creating a metadict class inheriting from dict ?).  I think this can be moved to a second PR.

keflavich commented:
I agree that the metadata handling can be moved to another PR.  I like the idea of handling metadata validity checking in a dictionary subclass.

keflavich commented on regions/io/crtf/read.py:
```diff
@@ -29,7 +29,7 @@
 
 # Extracts each 'parameter=value' pair.
-regex_meta = re.compile(r'((?:\w+\s*=\s*[^,\[\]]+)|(?:\w+\s*=\s*\[.*?\]))')
+regex_meta = re.compile(r'(?:(\w+)\s*=[\s\'\"]*([^,\[\]]+?)[\'\",]+)|(?:(\w+)\s*=\s*\[(.*?)\])')
```
if at all possible, add a comment describing this regex and/or indicating some strings it should match.  These should also be included in tests, of course, but regex's are really hard to read in general

sushobhana replied:
ya.. I will try to convert them into `VERBOSE` mode (there is a option to describe what you are doing in the regex)

migueldvb commented on regions/io/ds9/read.py:
```diff
@@ -5,42 +5,38 @@
 import re
```
Apart from changing things into `io.core`, are the other updates in `io.ds9` required in this pull request?

sushobhana replied:
Some of them are required, there have been some minor changes in the shape class. There are also some doc fixes which could have gone in a separate PR(My bad!)

migueldvb commented on docs/ds9.rst:
```diff
@@ -17,18 +17,19 @@ representing one DS9 region. The `~regions.Shape` objects can be converted to
     >>> parser = DS9Parser(reg_string)
```
Some commits have `[skip ci]` to bypass running the build. Just a reminder that if this goes in the commit message of the HEAD commit in the pull request the CI build will be passed over for this PR which is not what you want. I learned this by having made that mistake some time ago so better to use [skip ci] only for pull requests that do not need CI.

migueldvb replied:
Perhaps those fixes in ds9 could go in a separate pull request?

sushobhana replied:
Oh! I did not know about that!. I thought it's better to skip ci for minor doc fixes. I will take care of it from next time.
And about those those minor changes in ds9, most of them are related to the changes done in the io/core so I thought it's better to commit in this PR.

migueldvb commented:
Some tests are failing in Travis CI python 3.6 environment but apparently unrelated with these updates.

sushobhana commented:
The tests are failing due to some pytest deprecation. Should i be worried? @bsipocz

sushobhana commented:
@keflavich @migueldvb  It's just that I am making a lot of changes to Shape class  on top of the changes made here. Should I add those changes in this PR?( which is unrelated to converting CRTF strings to shape objects, )..

keflavich commented:
I'm not sure what you're asking.  Are you saying there are changes critical to this PR that you have not yet included?

I don't know what's going on with the travis failures; we either need to remove the deprecated code (which sounds like an upstream issue) or lock to a slightly older version of pytest.

sushobhana commented:
No, I have added methods for conversion of regions to shape and also refactored the whole `ds9_objects_to_string` method making them use of the shape layer , on top of the changes done to Shape class in this PR.  https://github.com/sushobhana/regions/tree/improvements.  Should I add those changes here and extend the scope of this PR ?

keflavich commented:
No, I think it's best to make that a new PR.  This one is nearly ready to merge, right?

sushobhana commented:
Ya, it's completely ready , just waiting for your signal!

migueldvb commented on regions/io/crtf/test/test_crtf_language.py:
```diff
@@ -0,0 +1,128 @@
+import os
```
`os` is imported but unused in the CRFT test.

migueldvb commented on regions/io/crtf/test/test_crtf_language.py:
```diff
@@ -0,0 +1,128 @@
+import os
+import distutils.version as vers
+
+from numpy.testing import assert_allclose
```
`assert_allclose` is apparently unused. Could you check the imports in the test files?

sushobhana replied:
sorry , I will remove the unused imports.

bsipocz commented:
@sushobhana @keflavich - The fix for the pytest related failures is coming with the new releases, ci-helpers will have the update in the next hour or so.
